src/txmempool.h

```diff
@@ -485,7 +485,18 @@ class CTxMemPool
         >
     > indexed_transaction_set;
 
-    mutable CCriticalSection cs;
+    /**
+     * This mutex needs to be locked when accessing mapTx or other members that
+     * are guarded by it.
+     *
+     * @par Relationship to cs_main
+     * Locking both mempool.cs and cs_main gives a view of the mempool which is
+     * compatible with the current chain tip. To guarantee this, both mutexes need
+     * to be locked when adding transactions to the mempool or changing the
+     * chain tip. But locking mempool.cs by itself is sufficient for reading
+     * transactions from the mempool, or removing transactions from it.
+     */
+    mutable RecursiveMutex cs;
```

Comment:
This comment comes from https://github.com/bitcoin/bitcoin/pull/14963#discussion_r242268194 and I think it is still ok as a description of how the code currently works. But talking to @morcos, @sdaftuar, and you Monday, it sounds like this is not actually how things are intended to work. After that conversation, I think the following might be more accurate:

```c++
/**
 * This mutex needs to be locked when accessing `mapTx` or other members
 * that are guarded by it.
 *
 * @par Consistency guarantees
 *
 * By design, it is guaranteed that:
 *
 * 1. Locking both `cs_main` and `mempool.cs` will give a view of mempool
 *    that is consistent with current chain tip (`chainActive` and
 *    `pcoinsTip`) and is fully populated. Fully populated means that if the
 *    current active chain is missing transactions that were present in a
 *    previously active chain, all the missing transactions will have been
 *    re-added to the mempool and should be present if they meet size and
 *    consistency constraints.
 *
 * 2. Locking `mempool.cs` without `cs_main` will give a view of a mempool
 *    consistent with some chain that was active since `cs_main` was last
 *    locked, and that is fully populated as described above. It is ok for
 *    code that only needs to query or remove transactions from the mempool
 *    to lock just `mempool.cs` without `cs_main`.
 *
 * To provide these guarantees, it is necessary to lock both `cs_main` and
 * `mempool.cs` whenever adding transactions to the mempool and whenever
 * changing the chain tip. It's necessary to keep both mutexes locked until
 * the mempool is consistent with the new chain tip and fully populated.
 *
 * @par Consistency bug
 *
 * The second guarantee above is not currently enforced, but
 * https://github.com/bitcoin/bitcoin/pull/14193 will fix it. No known code
 * in bitcoin currently depends on second guarantee, but it is important to
 * fix for third party code that needs be able to frequently poll the
 * mempool without locking `cs_main` and without encountering missing
 * transactions during reorgs.
 */
```

Reply:
Wow, excellent write-up. Will pull that in to replace my clumsy explanation